r/R/io.R

```diff
@@ -241,7 +241,9 @@ mmap_open <- function(path, mode = c("read", "write", "readwrite")) {
 make_readable_file <- function(file, mmap = TRUE, compression = NULL, filesystem = NULL) {
   if (inherits(file, "SubTreeFileSystem")) {
     filesystem <- file$base_fs
-    file <- file$base_path
+    # SubTreeFileSystem adds a slash to base_path, but filesystems will reject file names
+    # with trailing slashes, so we need to remove it here.
+    file <- sub("/$", "", file$base_path)
```

Comment:
I'm probably misreading this, but is this treating any `SubTreeFileSystem` as pointing to a local file path?

Reply:
By "filesystems" I meant the Arrow FileSystem classes, not the local file system. Does that clarify your confusion? Or are you asking something else?

Reply:
No, I am asking something else. It seems that this code is replacing `file` (a FileSystem instance) with a file path, is that right? And below, the file path `file` will be treated as a local filesystem path?

Reply:
But the path should be treated as correct filesystem since we extracted that out in the line before.

Reply:
Oh, I see, I had missed the `filesystem <- file$base_fs`. So I guess my last question is: why are we calling `make_readable_file` with a `SubTreeFileSystem` as the first argument?

Reply:
The expectation is that users will call `FileSystem$path()` to specify the location to write/read to:

```R
fs <- S3FileSystem$create()
write_parquet(my_tab, fs$path("my/path/to"))
```

`fs$path` returns a `SubTreeFileSystem` as a convenient way to bundle a path and filesystem object in a single object. See discussion: https://github.com/apache/arrow/pull/8351#discussion_r501252247

Reply:
Ah. That's unfortunate, but that's not this PR's business. Thanks for the details!